crates/qasm2/src/lib.rs

```diff
@@ -0,0 +1,137 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2023
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+#[macro_use]
+extern crate lazy_static;
+
+use pyo3::prelude::*;
+use pyo3::Python;
+
+mod bytecode;
+mod error;
+mod expr;
+mod lex;
+mod parse;
+
+/// Information about a custom instruction that Python space is able to construct to pass down to
+/// us.
+#[pyclass]
+#[derive(Clone)]
+pub struct CustomInstruction {
+    pub name: String,
+    pub num_params: usize,
+    pub num_qubits: usize,
+    pub builtin: bool,
+}
+
+#[pymethods]
+impl CustomInstruction {
+    #[new]
+    fn __new__(name: String, num_params: usize, num_qubits: usize, builtin: bool) -> Self {
+        Self {
+            name,
+            num_params,
+            num_qubits,
+            builtin,
+        }
+    }
+}
+
+/// Information about a custom classical function that should be defined in mathematical
+/// expressions.
+///
+/// The given `callable` must be a Python function that takes `num_params` floats, and returns a
+/// float.  The `name` is the identifier that refers to it in the OpenQASM 2 program.  This cannot
+/// clash with any defined gates.
+#[pyclass(text_signature = "(name, num_params, callable, /)")]
+#[derive(Clone)]
+pub struct CustomClassical {
+    pub name: String,
+    pub num_params: usize,
+    pub callable: PyObject,
+}
+
+#[pymethods]
+impl CustomClassical {
+    #[new]
+    fn __new__(name: String, num_params: usize, callable: PyObject) -> Self {
+        Self {
+            name,
+            num_params,
+            callable,
+        }
+    }
+}
+
+/// Create a bytecode iterable from a string containing an OpenQASM 2 program.  The iterable will
+/// lex and parse the source lazily; evaluating OpenQASM 2 statements as required, without loading
+/// the entire token and parse tree into memory at once.
+#[pyfunction]
+fn bytecode_from_string(
+    _py: Python<'_>,
+    string: String,
+    include_path: Vec<std::ffi::OsString>,
```

Comment:
Btw you can use `PathBuf` with PyO3, and it will work if the user passes a Python `str`:

https://pyo3.rs/v0.18.1/conversions/tables#argument-types

Reply:
Ah cool, thanks.  Fixed in f74c87204acbd38a24a43c47d3da75361b617335.  I left the `filename` of the various lexer/parser state objects as `OsString` because being a diagnostic thing only, those don't actually _need_ to be paths (e.g. for string-loading mode, they're set to `<input>`).